index.bs

```diff
@@ -12349,12 +12363,60 @@ It is the responsibility of specifications using Web IDL to state
 which global environment (or, by proxy, which global object) each platform
 object is associated with.
 
+<div algorithm>
+  To <dfn export lt="new|create|create a new|create an object implementing|create an object
+  implementing the interface">create an object implementing the interface</dfn> |interface| in the
+  Realm |realm|, with optional `new.target` value |newTarget|, perform the following steps:
+
+    1.  Assert: |interface| is [=exposed=] in |realm|.
+    1.  Let prototype be the [=interface prototype object=] for |interface| in
+        |realm|.
+    1.  If |newTarget| is provided, then:
+        1.  Issue: If we just pass {{NewTarget}} through in a constructor, would this not
+            always be defined (sometimes to the [=interface object=] itself)?
```

Comment:
I'm not sure if it makes sense to take _newTarget_ as a parameter like this, as opposed to somehow retrieving it from the ambient algorithm. If we do that, then I imagine we'd have _newTarget_ provided if this algorithm is being invoked from a constructor, and we wouldn't have it provided if it's being invoked from some other algorithm.

Reply:
If I am understanding @littledan correctly, +1. Inside a constructor "Create a new X" should implicitly grab the NewTarget that exists. Elsewhere, it should just be the X constructor. In neither case should the caller of "new" pass anything.

(In the future I think constructors would be nicer if Web IDL created the object for you, and constructor spec text elsewhere just operated on the created object using a known variable name like `this` or something. But, that's a bigger project, and maybe not worth the churn. For now we have to work with the world where all constructors say "create a new X" as their first step or so.)

Reply:
I guess we should also make the Realm optional inside constructors; it should always be the current Realm, and asking people to figure that out is a recipe for badness.

Reply:
(I think even if we improved how to define a constructor, you still need new, e.g., for tear-off objects and such.)

Reply:
That would have some extremely surprising effects if interpreted to the letter. Consider the [`Response` constructor](https://fetch.spec.whatwg.org/#dom-response):

3. Let _r_ be a new `Response` object associated with a new response.
4. Set _r_’s headers to a new `Headers` object, whose header list is _r_’s response’s header list, and guard is "response".

If you just pull the NewTarget from the air, it would be used for both the `Response` and `Headers` objects.

Reply:
Yes, I guess you'd need to guard it to only apply to the creation of objects of the interface whose constructor is being interpreted.

Reply:
I still think that pulling the NewTarget from the ambient algorithm makes this unnecessarily hard to follow. Alternative proposal: when passing a NewTarget, we derive the realm from that. That at least means that specs defining constructors don't need to think about realms there.

Reply:
I don't quite understand the alternate proposal. Can you explain what would be required of spec authors using the "new" construct in your version?

From my perspective, I'm not very concerned with how hard to follow this algorithm is. I'm concerned about users of this algorithm, and ensuring they don't have to pass any explicit information in constructors, and only have to pass one thing (the realm) in other contexts.

Reply:
In the interest of making life easy for the users of the algorithm, I've pushed a change to do the object creation in WebIDL. If we're going to edit all the specifications that create objects anyway, we might as well make take the opportunity to simplify it as much as possible.